npm/pkg/dataplane/dpshim/dpshim.go

```diff
@@ -1,77 +1,590 @@
+		prefixedListName := listMetadata.GetPrefixName()
+		if !dp.setExists(prefixedListName) {
+			dp.createIPSet(listMetadata)
+		}
+
+		list := dp.setCache[prefixedListName]
+		if list.IPSetMetadata.GetSetKind() != ipsets.ListSet {
+			return npmerrors.Errorf(npmErrorString, false, fmt.Sprintf("ipset %s is not a list set", prefixedListName))
+		}
+	}
+
+	for _, memberMetadata := range memberMetadatas {
+		memberName := memberMetadata.GetPrefixName()
+		if !dp.setExists(memberName) {
+			dp.createIPSet(memberMetadata)
+		}
+		member := dp.setCache[memberName]
+
+		// Nested IPSets are only supported for windows
+		// Check if we want to actually use that support
+		if member.IPSetMetadata.GetSetKind() != ipsets.HashSet {
+			return npmerrors.Errorf(npmErrorString, false, fmt.Sprintf("ipset %s is not a hash set and nested list sets are not supported", memberName))
+		}
+	}
+	return nil
+}
+
+func (dp *DPShim) processIPSetsApply() (*protos.GoalState, error) {
+	if len(dp.dirtyCache.toAddorUpdateSets) == 0 {
+		return nil, nil
+	}
+
+	toApplySets := make([]*controlplane.ControllerIPSets, len(dp.dirtyCache.toAddorUpdateSets))
+	idx := 0
+
+	for setName := range dp.dirtyCache.toAddorUpdateSets {
+		set := dp.getIPSet(setName)
+		if set == nil {
+			klog.Errorf("processIPSetsApply: set %s not found", setName)
+			return nil, npmerrors.Errorf(npmerrors.AppendIPSet, false, fmt.Sprintf("ipset %s not found", setName))
+		}
+
+		toApplySets[idx] = set
+		idx++
+	}
+
+	payload, err := controlplane.EncodeControllerIPSets(toApplySets)
+	if err != nil {
+		klog.Errorf("processIPSetsApply: failed to encode sets %v", err)
+		return nil, npmerrors.ErrorWrapper(npmerrors.AppendIPSet, false, "processIPSetsApply: failed to encode sets", err)
+	}
+
+	return getGoalStateFromBuffer(payload), nil
+}
+
+func (dp *DPShim) processIPSetsDelete() (*protos.GoalState, error) {
+	if len(dp.dirtyCache.toDeleteSets) == 0 {
+		return nil, nil
+	}
+
+	toDeleteSets := make([]string, len(dp.dirtyCache.toDeleteSets))
+	idx := 0
+
+	for setName := range dp.dirtyCache.toDeleteSets {
+		toDeleteSets[idx] = setName
+		idx++
+	}
+
+	payload, err := controlplane.EncodeStrings(toDeleteSets)
+	if err != nil {
+		klog.Errorf("processIPSetsDelete: failed to encode sets %v", err)
+		return nil, npmerrors.ErrorWrapper(npmerrors.DeleteIPSet, false, "processIPSetsDelete: failed to encode sets", err)
+	}
+
+	return getGoalStateFromBuffer(payload), nil
+}
+
+func (dp *DPShim) processPoliciesApply() (*protos.GoalState, error) {
+	if len(dp.dirtyCache.toAddorUpdatePolicies) == 0 {
+		return nil, nil
+	}
+
+	toApplyPolicies := make([]*policies.NPMNetworkPolicy, len(dp.dirtyCache.toAddorUpdatePolicies))
+	idx := 0
+
+	for policyKey := range dp.dirtyCache.toAddorUpdatePolicies {
+		if !dp.policyExists(policyKey) {
+			return nil, npmerrors.Errorf(npmerrors.AddPolicy, false, fmt.Sprintf("policy %s not found", policyKey))
+		}
+
+		policy := dp.policyCache[policyKey]
```

Comment:
What if the policy doesnt exist?

Reply:
Thats the check we are doing one line above ?